tcmalloc/internal/percpu_rseq_aarch64.S

```diff
@@ -0,0 +1,487 @@
+  mrs	  x5, tpidr_el0;			\
+  add	  x5, x5, x0;				\
+  ldp	  x2, x3, [sp, #16];			\
+  ldp	  x0, x1, [sp], #48;			\
+  adrp	  x6, __rseq_cs_##src;			\
+  add     x6, x6, :lo12:__rseq_cs_##src;	\
+  str     x6, [x5, #8];				\
+  .L##src##_start:
+
+#endif
+/* ---------------- end helper macros ---------------- */
+
+/* start of atomic restartable sequences */
+
+/*
+ * int TcmallocSlab_PerCpuCmpxchg64(int target_cpu, long *p,
+ *                                  long old_val, long new_val)
+ * w0: target_cpu
+ * x1: p
+ * x2: old_val
+ * x3: new_val
+ */
+  .p2align 6 /* aligns to 2^6 with NOP filling */
+  .globl TcmallocSlab_PerCpuCmpxchg64
+  .type  TcmallocSlab_PerCpuCmpxchg64, @function
+TcmallocSlab_PerCpuCmpxchg64:
+  .cfi_startproc
+.LTcmallocSlab_PerCpuCmpxchg64_region0:
+  START_RSEQ(TcmallocSlab_PerCpuCmpxchg64)
+  FETCH_CPU(w4)
+  cmp w0, w4 /* check cpu vs current_cpu */
+  bne .LTcmallocSlab_PerCpuCmpxchg64_region1
+  ldr x6, [x1]
+  cmp x6, x2 /* verify *p == old */
+  bne .LTcmallocSlab_PerCpuCmpxchg64_region2
+  str x3, [x1]
+.LTcmallocSlab_PerCpuCmpxchg64_region1:
+.LTcmallocSlab_PerCpuCmpxchg64_commit:
+  mov x0, x4
+  ret  /* return current cpu, indicating mismatch OR success */
+.LTcmallocSlab_PerCpuCmpxchg64_region2:
+  mov x0, #-1 /* mismatch versus "old" or "check", return -1 */
+  ret
+.LTcmallocSlab_PerCpuCmpxchg64_region3:
+  .cfi_endproc
+ENCODE_SIZE(TcmallocSlab_PerCpuCmpxchg64)
+DEFINE_UPSTREAM_CS(TcmallocSlab_PerCpuCmpxchg64)
+
+/* size_t TcmallocSlab_PushBatch_FixedShift(
+ *     void *ptr (x0),
+ *     size_t cl (w1),
+ *     void** batch (x2),
+ *     size_t len (w3) {
+ *   uint64_t x8 = __rseq_abi.cpu_id
+ *   uint64_t* x8 = CpuMemoryStart(x0, r8)
+ *   Header* hdr = x8 + w1 * 8
+ *   uint64_t x9 = hdr->current (zero-extend 16bit)
+ *   uint64_t x10 = hdr->end	(zero-extend 16bit)
+ *   if (w9 >= w10) return 0
+ *   x11 = x3
+ *   w10 = w9 + min(w3, w10 - w9)
+ * loop:
+ *   x11--
+ *   x5 = *(batch + x11 * 8)
+ *   *(x8 + x9 * 8) = x5
+ *   x9++
+ *   if (x9 != x10) goto loop
+ *   hdr->current = r9 (16bit store)
+ *   return len - r11
+ * }
+ */
+  .p2align 6 /* aligns to 2^6 with NOP filling */
+  .globl TcmallocSlab_PushBatch_FixedShift
+  .type  TcmallocSlab_PushBatch_FixedShift, @function
+TcmallocSlab_PushBatch_FixedShift:
+  .cfi_startproc
+.LTcmallocSlab_PushBatch_FixedShift_region0:
+  START_RSEQ(TcmallocSlab_PushBatch_FixedShift)
+  FETCH_CPU(w8)
+  lsl x8, x8, #PERCPU_TCMALLOC_FIXED_SLAB_SHIFT /* multiply cpu by 256k */
+  add x8, x0, x8
+  add x4, x8, x1, LSL #3
+  ldrh w9, [x4] /* current */
+  ldrh w10, [x4, #6] /* end */
+  cmp w9, w10
+  bge .LTcmallocSlab_PushBatch_FixedShift_region2
+  mov  x11, x3      /* r11 = copy of len */
+  sub  w10, w10, w9 /* r10 = free capacity */
+  cmp w11, w10
+  csel w10, w11, w10, ls /* r10 = min(len, free capacity) */
+  add x10, x10, x9
+.LTcmallocSlab_PushBatch_FixedShift_loop:
+  sub x11 ,x11, #1
+  ldr x12, [x2, x11, LSL #3]
+  str x12, [x8, x9, LSL #3]
```

Comment:
I think you could save two instructions in the loop here by pre-shifting `x9` and `x11`, and then using pre-index addressing in the loop.

Reply:
As I was writing this I also thought I might be able to optimize the accesses here.

I just had a quick look though and it does not look that simple unfortunately. Pre-index takes a single register and positive immediate. So traversing the Batch in reverse order wouldn't work with pre-index. Also the store would be problematic, as we would need to precompute x8+x9 >> 3, then increase by 8. But x9 is used to compare to x10, which holds the value of x9 (current) + amount of items beeing pushed. We could add x8 to both, and compare them but that's an extra instruction, though both would be outside the loop, might be worth looking at.

Reply:
Oh and to work around the negative offseting there is LDUR, but AFAICT that has no writeback variant.

Reply:
Are you sure that negative pre-index addressing isn't allowed?  My copy of LLVM happily accepts an instruction of the form `ldr	x1, [x2, #-4]!`, and my copy of the ISA reference shows that the immediate operand is sign-extended even in the pre-indexed form of the instruction.

Reply:
Yeah you are right, my eyes seem to have skipped that entire part. I will have a look at using these other addressing modes.

Thanks!

Reply:
I have updated the PushBatch and PopBatch to reduce the number of instructions in the loop. I had a look at the the other Push/Pop sequences and I don't think those would benefit from hoisting out the shift from the load, since you need the +1 anyway to store back the current index.

The only way to improve this would be to keep the current, begin and end offsets as multiples of 8. Though I suspect these are used elsewhere and making this change isn't that straighforward.